gateway/httpproxy/httpproxy.go

```diff
@@ -0,0 +1,117 @@
+package httpproxy
+
+import (
+	"bufio"
+	"io"
+	"net"
+	"net/http"
+	"net/textproto"
+	"strconv"
+	"strings"
+
+	"github.com/nknorg/nkn/util/config"
+	"github.com/nknorg/nkn/util/log"
+	"github.com/xtaci/smux"
+)
+
+type HTTPProxy struct {
+	// defines a listeners for http proxy, such as "127.0.0.1:30004"
+	listener string
+}
+
+func NewServer() *HTTPProxy {
+	return &HTTPProxy{
+		listener: ":" + strconv.Itoa(int(config.Parameters.HttpProxyPort)),
+	}
+}
+
+func transfer(destination io.WriteCloser, source io.ReadCloser) {
+	defer destination.Close()
+	defer source.Close()
+	io.Copy(destination, source)
+}
+
+// parseRequestLine parses "GET /foo HTTP/1.1" into its three parts.
+func parseRequestLine(line string) (method, requestURI, proto string, ok bool) {
+	s1 := strings.Index(line, " ")
+	s2 := strings.Index(line[s1+1:], " ")
+	if s1 < 0 || s2 < 0 {
+		return
+	}
+	s2 += s1 + 1
+	return line[:s1], line[s1+1 : s2], line[s2+1:], true
+}
+
+func closeConnection(conn net.Conn) {
+	err := conn.Close()
+	if err != nil {
+		log.Error("Error while closing connection:", err)
+	}
+}
+
+func handleSession(conn net.Conn, session *smux.Session) {
+	for {
+		stream, err := session.AcceptStream()
+		if err != nil {
+			log.Error("Couldn't accept stream:", err)
+			break
+		}
+
+		tp := textproto.NewReader(bufio.NewReader(stream))
+
+		line, err := tp.ReadLine()
+		if err != nil {
+			log.Error("Couldn't read line:", err)
+			break
+		}
+
+		method, host, _, _ := parseRequestLine(line)
+
+		// won't proxy HTTP due to security reasons
+		if method != http.MethodConnect {
+			log.Error("Only CONNECT HTTP method supported")
+			stream.Write([]byte("HTTP/1.1 403 Forbidden\r\n\r\n"))
+			stream.Close()
+			break
+		}
+
+		destConn, err := net.Dial("tcp", host) //TODO: add timeout
+		if err != nil {
+			log.Error("Couldn't connect to host", host, "with error:", err)
+			break
+		}
+
+		stream.Write([]byte("HTTP/1.1 200 OK\r\n\r\n"))
+
+		go transfer(destConn, stream)
+		go transfer(stream, destConn)
+	}
+
+	session.Close()
+	conn.Close()
```

Comment:
why? they're called each time `break` is called in loop above

Reply:
I think the current "problem" are:
1. if a stream get wrong, then all steams are closed (since session and conn are closed)
2. if non of the streams get wrong, then all streams or session or conn are never closed

Reply:
1. session is limited to one client, so it there is some error then (I think) it's better to recreate session/streams just in case (client software will handle this transparently)
2. it will be called if `AcceptStream` will return error, which will happen when session or connection is closed from client side

Reply:
This is what I think would be more robust:
1. If accept new stream error, then close conn
2. Once stream is accepted, if any error occurs to that stream, only close that stream so that other stream will not be affected.

I think the use case is that, user will open a client as daemon, all apps will use the client as proxy. If any application make some mistake, other application shouldn't be affected. Actually this is very similar to kcptun (https://github.com/xtaci/kcptun/ ) and they use the above logic.

Reply:
I think I already did what you suggested in the latest commit, please take a look

Reply:
You are right :)